scheduler/reconcile.go

```diff
@@ -778,14 +688,177 @@ func (a *allocReconciler) computePlacements(group *structs.TaskGroup,
 			place = append(place, allocPlaceResult{
 				name:               name,
 				taskGroup:          group,
-				downgradeNonCanary: canaryState,
+				downgradeNonCanary: isCanarying,
 			})
 		}
 	}
 
 	return place
 }
 
+// computeReplacements either applies the placements calculated by computePlacements,
+// or computes more placements based on whether the deployment is ready for placement
+// and if the placement is already rescheduling or part of a failed deployment.
+// The input deploymentPlaceReady is calculated as the deployment is not paused, failed, or canarying.
+// It returns the number of allocs still needed.
+func (a *allocReconciler) computeReplacements(deploymentPlaceReady bool, desiredChanges *structs.DesiredUpdates,
+	place []allocPlaceResult, failed, lost allocSet, underProvisionedBy int) int {
+
+	// If the deployment is place ready, apply all placements and return
+	if deploymentPlaceReady {
+		desiredChanges.Place += uint64(len(place))
+		// This relies on the computePlacements having built this set, which in
+		// turn relies on len(lostLater) == 0.
+		a.result.place = append(a.result.place, place...)
+		a.markStop(failed, "", allocRescheduled)
+		desiredChanges.Stop += uint64(len(failed))
+
+		min := helper.IntMin(len(place), underProvisionedBy)
+		underProvisionedBy -= min
+		return underProvisionedBy
+	}
+
+	// We do not want to place additional allocations but in the case we
+	// have lost allocations or allocations that require rescheduling now,
+	// we do so regardless to avoid odd user experiences.
+
+	// If allocs have been lost, determine the number of replacements that are needed
+	// and add placements to the result for the lost allocs.
+	if len(lost) != 0 {
+		allowed := helper.IntMin(len(lost), len(place))
+		desiredChanges.Place += uint64(allowed)
+		a.result.place = append(a.result.place, place[:allowed]...)
+	}
+
+	// if no failures or there are no pending placements return.
+	if len(failed) == 0 || len(place) == 0 {
```

Comment:
It's probably fine to leave in for intent, but the second half of the conditional doesn't get us anything, as we'll be merrily skipping past the next section if `len(place)` is empty.